### PR TITLE
Pluggable Index Templates

### PR DESCRIPTION
**Pull request by diesaulov**
Due to the hard-coded nature of the Index Template type, the core server module requires changes every time a new custom Index Template is needed to build a new feature, what violates the Open-Closed principle and makes development less flexible.

This PR replaces the `TemplateType` enum with a string value and adds support of external Index Template providers.

/jenkins-pr-deps https://github.com/Graylog2/graylog-plugin-enterprise/pull/2702

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/shared/bindings/GenericBindings.java [hunk 1]
```diff
@@ -90,6 +94,12 @@ protected void configure() {
                 .setDefault()
                 .to(DefaultFailureHandlingConfiguration.class);
 
+        final Multibinder<IndexTemplateProvider> indexTemplateProviderMultibinder = Multibinder.newSetBinder(binder(), IndexTemplateProvider.class);
+
+        indexTemplateProviderMultibinder.addBinding().to(MessageIndexTemplateProvider.class);
+        indexTemplateProviderMultibinder.addBinding().to(EventIndexTemplateProvider.class);
+        indexTemplateProviderMultibinder.addBinding().to(GimIndexTemplateProvider.class);
+
```
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/shared/bindings/GenericBindings.java:102]: We usually use a map binder and use the type string as key. Using a set binder and then find factories by iterating over the set and comparing the type works as well, of course. Any specific reason why you didn't choose a map binder?

Apart from the easier lookup, another advantage of the map binder is, that trying to bind the same key twice results in a binding error on server startup. With the set binder we have to check that there is only a single implementation for the same key at runtime and throw an IllegalStateException.

**Review by bernd — CHANGES_REQUESTED** [review]
Thanks for the refactoring! :pray: I will test this tomorrow. One thing I noticed:

File: graylog2-server/src/main/java/org/graylog2/shared/bindings/GenericBindings.java (hunk 1 above)
Comment (diesaulov) [reply graylog2-server/src/main/java/org/graylog2/shared/bindings/GenericBindings.java]: https://github.com/Graylog2/graylog2-server/pull/11290/commits/abcd797052c00b60b2e09380fa5085ed706e14f8

**Review by bernd — APPROVED** [review]
Tested in my dev environment and LGTM! :+1: Good work, thank you!

I think at some point in the future we can add an `addIndexTemplateType(String, Class<? extends IndexTemplateProvider>)` to `PluginModule` like we do for other pluggable objects to make it easier for plugins to register them.

**Pull request merged**